File: assignments/assignment-1-solution/src/CrazyEights/Player/CpuPlayer.cs
using CrazyEights.Cards;
using CrazyEights.Domain;
using CrazyEights.Game;

namespace CrazyEights.Player;

public class CpuPlayer : PlayerBase
{
    private const string CpuPlayerName = "CPU";

    public CpuPlayer(Hand hand, bool showHand) : base(CpuPlayerName, hand, showHand)
    {
    }

    public override ICard SelectCard(TurnContext turnContext)
    {
        var rng = turnContext.RandomNumberGenerator;
        var playableCards = Hand.GetPlayableCards(turnContext.CurrentSuit, turnContext.CurrentRank);

        //cannot play a card
        if (playableCards.Count == 0) return UnselectedCard.Instance;

        //select whether to play a card or not
        var playACard = rng.NextDouble() > 0.5;
        if (!playACard) return UnselectedCard.Instance;

        //randomly select a card to play
        var randomCardIndex =
            turnContext.RandomNumberGenerator.Next(0, playableCards.Count);
        var selectedCard = playableCards[randomCardIndex].Card;

        return selectedCard;
    }

    public override SuitType SelectSuit(GameContext gameContext, TurnContext turnContext)
    {
        var rng = turnContext.RandomNumberGenerator;
        var suitValues = Suit.GetSuits();
        var randomSuitIndex = rng.Next(0, suitValues.Count - 1);
        var selectedSuit = suitValues[randomSuitIndex];
        return selectedSuit;
    }

    public override bool WillPlayDrawnCard(TurnContext turnContext, ICard drawnCard)
    {
        var rng = turnContext.RandomNumberGenerator;
        var playDrawnCard = rng.NextDouble() > 0.5;
        return playDrawnCard;
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/player/CpuPlayer.java
package edu.kennesaw.crazy8s.player;

import edu.kennesaw.crazy8s.cards.Card;
import edu.kennesaw.crazy8s.cards.UnselectedCard;
import edu.kennesaw.crazy8s.domain.PlayableCard;
import edu.kennesaw.crazy8s.domain.Suit;
import edu.kennesaw.crazy8s.domain.SuitType;
import edu.kennesaw.crazy8s.game.GameContext;
import edu.kennesaw.crazy8s.game.TurnContext;
import java.util.List;
import java.util.Random;

/**
 * CPU-controlled player that makes random selections.
 */
public class CpuPlayer extends PlayerBase {
    private static final String CPU_PLAYER_NAME = "CPU";

    public CpuPlayer(Hand hand, boolean showHand) {
        super(CPU_PLAYER_NAME, hand, showHand);
    }

    @Override
    public Card selectCard(TurnContext turnContext) {
        Random rng = turnContext.getRandomNumberGenerator();
        List<PlayableCard> playableCards =
                getHand().getPlayableCards(turnContext.getCurrentSuit(), turnContext.getCurrentRank());

        if (playableCards.isEmpty()) {
            return UnselectedCard.getInstance();
        }

        int randomCardIndex = rng.nextInt(playableCards.size());
        return playableCards.get(randomCardIndex).getCard();
    }

    @Override
    public SuitType selectSuit(GameContext gameContext, TurnContext turnContext) {
        Random rng = turnContext.getRandomNumberGenerator();
        List<SuitType> suitValues = Suit.getSuits();
        int randomSuitIndex = rng.nextInt(suitValues.size());
        return suitValues.get(randomSuitIndex);
    }

    @Override
    public boolean willPlayDrawnCard(TurnContext turnContext, Card drawnCard) {
        Random rng = turnContext.getRandomNumberGenerator();
        return rng.nextDouble() > 0.5;
    }
}
